### ISSUE TITLE
twisted.web.http.Request.getClient  shouldn't use socket.gethostbyaddr

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2252]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#2252|
|Type|enhancement|
|Created|2006-12-01 00:36:10Z|
|Branch|https://github.com/twisted/twisted/tree/request-getclient-2252|

First off according to exarkun can block for arbitrary periods of time.  Second it doesn't support "::FFFF:127.0.0.1" addresses, which means it breaks when using IPv6 some of the time.

<details><summary>Searchable metadata</summary>

```
trac-id__2252 2252
type__enhancement enhancement
reporter__dreid dreid
priority__normal normal
milestone__ 
branch__branches_request_getclient_2252 branches/request-getclient-2252
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1164933370000000 1164933370000000
changetime__1404432653876515 1404432653876515
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

fwiw, it does work with ipv6 on some systems:

```
>>> socket.gethostbyaddr('::ffff:127.0.0.1')
('kunai', ['localhost.localdomain', 'localhost'], ['::ffff:127.0.0.1'])
>>> 
```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_2"></a>|@dreid commented|
|-|-|

To be more specific it works on some systems with some addresses.

On my edgy machine python 2.4.4c1

>>> socket.gethostbyaddr('::ffff:127.0.0.1')
('localhost', [], ['::ffff:127.0.0.1'])
>>> socket.gethostbyaddr('::ffff:209.157.133.61')
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
socket.error: Address family not supported by protocol

On my mac with python 2.4.4
>>> socket.gethostbyaddr('::ffff:127.0.0.1')
('localhost', [], ['127.0.0.1'])
>>> socket.gethostbyaddr('::ffff:209.157.133.61')
('christina.dreid.org', [], ['209.157.133.61'])

Such inconsistency is even more reason to not use it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I can't see any reason this needs to be targeted at 9.0 (8.2+1).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Not going to fix this in web2 (see [#4821](https://github.com/twisted/twisted/issues/4821)), but `twisted.web.http.Request.getClient` has the same problem.  Retargeting at web component.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [42835]) Branching to 'request-getclient-2252'

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set owner to @exarkun|
|-|-|

First, a general point:

1. I think this breaks the letter of the compatibility policy by changing the return value to None now instead of after two releases. People may be relying on this. I guess CompatibilityPolicy says we should talk about it on the mailing list. 

as for technical points:

2. there's a typo in the test_getClient method in the test_http.py file: "if favor" instead of "in favor".

3. minor: is there a reason you put the deprecated() assignment so far away from the getClient definition (instead of e.g. using the decorator syntax)?

4. Maybe requesthelper's getClient should also be deprecated? if tests are using that helper, the author of those tests may not realize that they're using deprecated APIs.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @radix|
|-|-|

Thanks for the review.

> I think this breaks the letter of the compatibility policy by changing the return value to None now instead of after two releases. People may be relying on this. I guess CompatibilityPolicy says we should talk about it on the mailing list.

https://twistedmatrix.com/pipermail/twisted-python/2014-July/028496.html

> there's a typo in the test_getClient method in the test_http.py file: "if favor" instead of "in favor".

r42842.

> minor: is there a reason you put the deprecated() assignment so far away from the getClient definition (instead of e.g. using the decorator syntax)?

Yep, [#4804](https://github.com/twisted/twisted/issues/4804).

> Maybe requesthelper's getClient should also be deprecated? if tests are using that helper, the author of those tests may not realize that they're using deprecated APIs

Good point.  r42841.

Re-assigning to you because https://github.com/twisted-infra/trac-config/issues/14

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_12"></a>|@radix set owner to @exarkun|
|-|-|

Thanks for the update.

5. You need to update the news snippet to mention the new behavior.

Feel free to merge after that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [42845]) update the news fragment

refs [#2252](https://github.com/twisted/twisted/issues/2252)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [42846]) Merge request-getclient-2252

Author: exarkun
Reviewer: radix
Fixes: [#2252](https://github.com/twisted/twisted/issues/2252)

Remove the blocking IP-to-hostname resolution from `twisted.web.http.request.getClient`
and the `REMOTE_HOST` variable from the CGI environment created by `twisted.web.twcgi`.
